main.py

```diff
@@ -0,0 +1,347 @@
+    thumbnail_name = ndb.StringProperty()
+    thumbnail_key = ndb.StringProperty()
+    date = ndb.DateTimeProperty(auto_now_add=True)
+    labels = ndb.StringProperty(repeated=True)
+    original_photo = ndb.StringProperty()
+
+
+class MainHandler(webapp2.RequestHandler):
+    '''Home/news feed page (notification listing).'''
+    def get(self):
+        # Fetch all notifications in reverse date order.
+        notifications = Notification.query().order(
+            -Notification.date).fetch(NUM_NOTIFICATIONS_TO_DISPLAY)
+        template_values = {'notifications': notifications}
+        template = jinja_environment.get_template("notifications.html")
+        # Write to the appropriate html file.
+        self.response.write(template.render(template_values))
+
+
+class PhotosHandler(webapp2.RequestHandler):
+    '''All photos page: displays thumbnails of all uploaded photos.'''
+    def get(self):
+        # Get thumbnail references from datastore in reverse date order.
+        thumbnail_references = ThumbnailReference.query().order(
+            -ThumbnailReference.date).fetch()
+        # Build dictionary of img_url of thumbnail to thumbnail_references.
+        thumbnails = collections.OrderedDict()
+        for thumbnail_reference in thumbnail_references:
+            img_url = get_thumbnail(thumbnail_reference.thumbnail_key)
+            thumbnails[img_url] = thumbnail_reference
+        template_values = {'thumbnails': thumbnails}
+        template = jinja_environment.get_template("photos.html")
+        # Write to appropriate html file.
+        self.response.write(template.render(template_values))
+
+
+class SearchHandler(webapp2.RequestHandler):
+    '''Search page: displays search bar and all
+    thumbnails applicable to entered search term.'''
+    def get(self):
+        # Get search_term entered by user.
+        search_term = self.request.get('search-term').lower()
+        references = ThumbnailReference.query().order(
+            -ThumbnailReference.date).fetch()
+        # Build dictionary of img_url of thumbnails to
+        # thumbnail_references that have the given label.
+        thumbnails = collections.OrderedDict()
+        for reference in references:
+            labels = reference.labels
+            if search_term in labels:
+                img_url = get_thumbnail(reference.thumbnail_key)
+                thumbnails[img_url] = reference
+        template_values = {'thumbnails': thumbnails}
+        template = jinja_environment.get_template("search.html")
+        # Write to appropriate html file.
+        self.response.write(template.render(template_values))
+
+
+class ReceiveMessage(webapp2.RequestHandler):
+    '''For receiving Cloud Pub/Sub push messages and performing
+    related logic.'''
+    def post(self):
+        logging.debug('Post body: {}'.format(self.request.body))
+        message = json.loads(urllib.unquote(self.request.body).rstrip('='))
```

Comment:
what's up with the `rstrip`? Look at our [other](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/appengine/flexible/pubsub/main.py#L62) push notification sample, it should be base64 encoded?

Reply:
`rstrip` came from this example: https://github.com/GoogleCloudPlatform/cloud-pubsub-samples-python/blob/master/appengine-push/main.py

Reply:
However, as it doesn't seem necessary, it has since been removed